### PR TITLE
Add UI test for progress on the server that is not on the client

### PR DESCRIPTION
**Pull request by joshlory**

### CODE REVIEW

File: dashboard/test/ui/step_definitions/steps.rb [hunk 1]
```diff
@@ -271,6 +271,19 @@ def replace_hostname(url)
...
+    And I reload the page
+    And I wait to see ".header_middle"
+    And I wait for 10 seconds
+    And element ".header_middle a.level_link:last" has class "#{test_result}"
+    Then I am on "http://studio.code.org/s/hourofcode"
+    And I wait to see ".user-stats-block"
+    And I wait for 10 seconds
+    And element ".user-stats-block .games:first a.level_link:last" has class "#{test_result}"
+  }
+end
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb:289]: I'm concerned that this step definition is misleading on a few levels:
- It's specific to the `hourofcode` script (hard-coded line 280); it would be easy for someone testing a different script to try and use this step, and be surprised when it didn't work.  Maybe put `hourofcode` in the name?
- It always checks the last level of the script (`a.level_link:last` selector) which is not implied by the step name.
- It navigates to a new page and leaves you there, which feels like a side effect given the step name.  Now I understand why you only use this as the very last step in a scenario.
- I'd prefer arbitrary 10-second waits to be replaced with waiting for something specific as much as possible; if not possible here, then a comment explaining what the 10 seconds is for would be helpful.

I'm not sure we need to make this fully generic yet, but if not I feel like the step name should do a little more to call out how specific this is.

File: dashboard/test/ui/features/hourOfCode_signedIn.feature [hunk 2]
```diff
@@ -61,16 +61,22 @@ Scenario: Failing at puzzle 6, refreshing puzzle 6, bubble should show up as att
...
-  And I wait for 2 seconds
-  And element ".user-stats-block .games:first a.level_link:last" has class "perfect"
+  And element ".header_middle a.level_link:last" has class "not_tried"
+  And I close the dialog
+  And I wait until element "#runButton" is visible
+  And I press "runButton"
+  And I wait to see ".modal"
+  Then I am on "http://studio.code.org/hoc/reset"
+  Then I am on "http://studio.code.org/hoc/20?noautoplay=true"
+  And validate progress for level is "attempted"
```
Comment (islemaster) [inline dashboard/test/ui/features/hourOfCode_signedIn.feature]: This scenario doesn't read sufficiently plain-English yet.  Part of this is a lack of experience on my part - I don't know how these steps verify progress on the server that is not on the client.  It'd be nice if the test described that well.

File: dashboard/test/ui/features/hourOfCode_signedIn.feature [hunk 3]
```diff
@@ -61,16 +61,22 @@ Scenario: Failing at puzzle 6, refreshing puzzle 6, bubble should show up as att
...
 Scenario: Async progress write followed by a stale read
   Given I am on "http://studio.code.org/hoc/20?noautoplay=true"
-  And element ".header_middle a.level_link:first" does not have class "perfect"
+  And I wait to see ".header_middle"
+  And element ".header_middle a.level_link:last" does not have class "perfect"
   Then mark the current level as completed on the client
-  And I reload the page
+  And validate progress for level is "perfect"
+
+Scenario: Progress on the server that is not on the client
```
Comment (islemaster) [inline dashboard/test/ui/features/hourOfCode_signedIn.feature:65]: Is this scenario name a precondition or a desired outcome? I can't tell.  Maybe rephrase to be more explicit about the expected behavior.

File: dashboard/test/ui/step_definitions/steps.rb [hunk 4]
```diff
@@ -271,6 +271,19 @@ def replace_hostname(url)
   @browser.execute_script %q-sessionStorage.setItem('progress', '{"hourofcode":{"' + appOptions.serverLevelId + '":100}}')-
 end
 
+Then /^validate progress for level is "([^"]*)"/ do |test_result|
+  steps %{
+    And I reload the page
+    And I wait to see ".header_middle"
+    And I wait for 10 seconds
+    And element ".header_middle a.level_link:last" has class "#{test_result}"
+    Then I am on "http://studio.code.org/s/hourofcode"
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb]: Suggestion: Break the second part of this (validating level progress on the script overview page) into its own step.  Then you can reuse the "validate progress for level" part in more places in the scenario, and be explicit about the "validate progress on script overview" as the last step in the scenario.

File: dashboard/test/ui/features/hourOfCode_signedIn.feature (hunk 2 above)
Comment (joshlory) [inline dashboard/test/ui/features/hourOfCode_signedIn.feature]: Good call, need to parameterize script and level number.

**Comment by joshlory** [comment]
```
./runner.rb -d localhost-studio.code.org:3000 --html -c ChromeLatestWin7 -f features/hourOfCode_signedIn.feature
Starting UI tests for ChromeLatestWin7_hourOfCode_signedIn
cucumber features/hourOfCode_signedIn.feature -t ~@eyes -t ~@eyes_mobile -t ~@local_only -t ~@skip -t ~@webpurify -t ~@pegasus_db_access -S --format html --out ChromeLatestWin7_hourOfCode_signedIn_output.html -f pretty
UI tests for ChromeLatestWin7_hourOfCode_signedIn succeeded (17:50 minutes, 6 passed)
```

```
./runner.rb -d localhost-studio.code.org:3000 --html -c ChromeLatestWin7 -f features/hourOfCode.feature 
Starting UI tests for ChromeLatestWin7_hourOfCode
cucumber features/hourOfCode.feature -t ~@eyes -t ~@eyes_mobile -t ~@local_only -t ~@skip -t ~@webpurify -t ~@pegasus_db_access -S --format html --out ChromeLatestWin7_hourOfCode_output.html -f pretty
UI tests for ChromeLatestWin7_hourOfCode succeeded (10:35 minutes, 4 passed)
```

File: dashboard/test/ui/step_definitions/steps.rb [hunk 5]
```diff
@@ -271,6 +271,23 @@ def replace_hostname(url)
...
+    And I wait for 10 seconds
+    And element ".progress_container a.level_link:nth(#{level.to_i - 1})" has class "#{test_result}"
+  }
+end
+
+Then /^I navigate to the course page and verify progress for course "([^"]*)" stage (\d+) level (\d+) is "([^"]*)"/ do |course, stage, level, test_result|
+  steps %{
+    Then I am on "http://studio.code.org/s/#{course}"
+    And I wait to see ".user-stats-block"
+    And I wait for 10 seconds
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb:286]: What's the purpose of the 10-second wait again?

File: dashboard/test/ui/features/hourOfCode_signedIn.feature [hunk 6]
```diff
@@ -57,20 +52,27 @@ Scenario: Failing at puzzle 6, refreshing puzzle 6, bubble should show up as att
...
+  And I navigate to the course page and verify progress for course "hourofcode" stage 1 level 20 is "perfect"
+
+Scenario: Progress on the server that is not on the client
+  Given I am on "http://studio.code.org/hoc/20?noautoplay=true"
+  And I verify progress in the header of the current page is "not_tried" for level 20
+  And I close the dialog
+  And I wait until element "#runButton" is visible
+  And I press "runButton"
+  And I wait to see ".modal"
+  Then I am on "http://studio.code.org/hoc/reset"
```
Comment (islemaster) [inline dashboard/test/ui/features/hourOfCode_signedIn.feature:72]: Nit: I'd still like a comment or something to the effect that this navigation step resets progress in local storage, but we still expect to get progress from the server.

**Comment by islemaster** [comment]
This is way, way better!  Looks great, I think you could merge this and address nits afterward if you like.

File: dashboard/test/ui/step_definitions/steps.rb (hunk 5 above)
Comment (joshlory) [inline dashboard/test/ui/step_definitions/steps.rb:286]: Allow time for the AJAX request to return in scenarios where progress is loaded async.
